### PR TITLE
Synteny modify

### PR DESCRIPTION
**Pull request by mschecht**
@ekiefl and @meren, I expanded `synteny.py` to accept a pangenome database which now allows the tool to create Ngrams with gene-clusters as apposed to gene-caller-ids or gene annotations. Ngrams are now extracted based on gene-cluster annotations, but the user is still reported the gene-caller-ids and annotation (selected by user) version of the Ngram. In the future, I want to refactor the tool to only use pangenome databases since I think this kind of analysis is intimately linked to pangenomics, but for now I think this is a decent solution.

Please have a look when you have a chance :)

### CODE REVIEW

File: anvio/synteny.py [hunk 1]
```diff
@@ -146,6 +152,10 @@ def sanity_check(self):
                 self.run.warning("Just so you know, %d contigs in %s had no genes. Here are the the first 5: %s" % \
                                   (len(contigs_without_genes), contigs_db_name, ''.join([str(x) for x in contigs_without_genes[:5]])))
 
+        #FIXME: Need a sanity check to test that a pan_db exists at the path given and is associated with the external genomes file
```
[the remark this answers is not in the document]
Comment (meren) [reply anvio/synteny.py:155]:
```
self.pan_db = PanDatabase(self.pan_db_path)
```

already performs this check. I think the point here is to make sure the external genomes file contains genome names that are compatible with the pan database. Almost requires a function like `utils.is_external_genomes_compatible_with_pan_database` :)

**Comment by meren** [comment]
> Since pan-dbs are now required input ...

But they shouldn't be :) And having them mandatory causes a lot of confusion of course. Here is the history behind the inclusion of pan dbs to synteny analyses to clarify things a bit:

Initially this program would get a single contigs database OR a list of them through ext/internal genomes files, and compute ngrams given a functional source. Functional source is key here, because ngrams are going to be occurrences of functions. That said, if a given gene did not have a function, it was a big problem since there is no way to count ngrams properly when the same known gene is associated with unknown genes in different contexts (think about a 2gram of `AX` and `AY` (where `A` is a gene with known function, `X` and `Y` are two genes with unknown functions)). The frequency of that 2gram would have been 2 and that may have been a mistake if we X and Y had different functions in reality.

Yet there is one solutoin: IF there is a pangenome that includes the contigs db OR list of contigs dbs of interest, this problem can be solved by annotating genes with unknown functions with the gene cluster to which they belong. The function is still unknown, but at least you know whether `X` = `Y` to avoid a mistake.

So pan databases should remain as an optional input to this program to make sense of genes with unknown functions.

**Comment by ekiefl** [comment]
I'll let @mschecht chime in, because that is not what we discussed during lab meeting once you left. I thought this was implemented because @mschecht was disappointed with the quality of ngrams created with COGs, even when each gene was assigned a COG.

**Comment by meren** [comment]
I see. I'll see what he has to add.

COGs may not be good enough, but the user may have a much better way to annotate their genes with functions. I think a flexible and user-friendly design for this concept would not _require_ a pan database for synteny analysis. In an ideal world, for a synteny analysis run on one or more contigs databases users should be able to,

- Use only on a single functional source (such as COGs, with the option to include or exclude genes with unknown functions)
- Use only on a pangenome to annotate genes with gene cluster names,
- Use a combination of a functional source and a pangenome to annotate genes that do not have any functions according to the functional source with gene cluster names.

For our own studies we can go with the second option, but how to design this for everyone is a different story :)

My 2 cents.

File: anvio/synteny.py [hunk 2]
```diff
@@ -154,12 +164,26 @@ def populate_genes(self):
         (ngram, count, contigs_db_name, contig_name, n)
 
         """
+        # Get gene cluster info from panDB
+        gene_cluster_frequencies_dataframe = pd.DataFrame.from_dict(
+                                                    self.pan_db.db.get_table_as_dict('gene_clusters'),
+                                                    orient='index')
```
[the remark this answers is not in the document]
Comment (mschecht) [reply anvio/synteny.py:170]: Thanks @ekiefl, just replaced it with the last line you suggested. I am still getting used to all the tool dbops has to extract great tables.

File: anvio/synteny.py [hunk 3]
```diff
@@ -193,12 +235,16 @@ def convert_to_df(self):
 
         for ngram_attribute in self.ngram_attributes_list:
             ngram = "::".join(map(str, list(ngram_attribute[0])))
-            df = pd.DataFrame(columns=['ngram', 'count', 'contig_db_name','contig_name','N','number_of_loci'])
+            ngram_function = "::".join(map(str, list(ngram_attribute[1])))
+            ngram_gene_clusters = "::".join(map(str, list(ngram_attribute[2])))
+            df = pd.DataFrame(columns=['ngram', 'ngram_functions', 'ngram_gene_clusters', 'count', 'contig_db_name','contig_name','N','number_of_loci'])
```
[the remark this answers is not in the document]
Comment (mschecht) [reply anvio/synteny.py:240]: Fixed, thanks!

File: anvio/dbops.py [hunk 4]
```diff
@@ -3421,7 +3421,7 @@ def __init__(self, db_path, run=run, progress=progress, quiet=True):
 
 
     def init(self):
-        if os.path.exists(self.db_path):
+        if filesnpaths.is_file_exists(self.db_path):
```
Comment (mschecht) [inline anvio/dbops.py:3424]: @meren I tried your suggestion in a previous comment use `self.pan_db = PanDatabase(self.pan_db_path)` to verify the PanDatabase path exists but when I tested it with a fake path no `ConfigError` was raised. Could this be a solution?

Comment (meren) [reply anvio/dbops.py:3424]: This would've been even better:

```
utils.is_pan_db(self.db_path)
```

**Comment by mschecht** [comment]
@ekiefl and @meren thank you very much for comments. Your suggestions push me to be a better programmer every time.

> 1. This is not an essential column (what if people just want gene cluster ngrams or gene caller id ngrams, or they only have a PAN.db because their colleague did not send over the 3000 contigs databases?). `--external-genomes` should really be optional. Unfortunately, this class has been initially designed around interacting with an `--external-genomes`, so I understand that the code would require significant refactoring to avoid this.

@ekiefl thank you for this suggestion. I tried out the idea and found that Ngrams made only gene-callers-id were not informative unless they are accompanied by some sort of annotation.  This is because Ngrams made of gene-callers-ids (e.g. 0::1::2) are found in every loci due to the contigsDB structure. I think the output of the tool needs to have either the columns `ngram_functions` or `ngram_gene_clusters` or both.

> 2. Since pan-dbs are now required input, functions should really be accessed through `--external-genomes` big brother, `--genomes-storage`. It is one succinct file (easy to pass around with collaborators, 1 table defines all functions amongst all genomes) and its required to generate the corresponding pan-db.

@ekiefl this is a great idea. I attempted to adjust this but found that, as you predicted, it requires a lot of refactoring. I would like to make this update in the future but if you think it is critical I'll get back on it. For now, I have added `FIXME`'s where I see potential changes needed to be made to address this. 

>  I think a flexible and user-friendly design for this concept would not _require_ a pan database for synteny analysis.

@meren I agree with this approach and am moving the pangenome parameter to optional.

> In an ideal world, for a synteny analysis run on one or more contigs databases users sho
> 
> * Use only on a single functional source (such as COGs, with the option to include or exclude genes with unknown functions)
> * Use only on a pangenome to annotate genes with gene cluster names,
> * Use a combination of a functional source and a pangenome to annotate genes that do not have any functions according to the functional source with gene cluster names.
> 
> For our own studies we can go with the second option, but how to design this for everyone is a different story :)

Thanks @meren this was super helpful to organize my ideas. I am attempting to makes flags so that the user has to input at least one of the annotation sources so there is maximum flexibility. If they provide more than one (e.g. pangenome and COGs) I will have a flag where the user will have to designate which annotation source should be used to actually slice up the ngrams. The other annotations will then be co-reported in the final datatable.

**Comment by mschecht** [comment]
@ekiefl @meren when you have a moment please look at my most recent commit. There is a functioning  test in `run_all_tests.sh`. @ekiefl's idea of using gene-callers-ids at the unit for which Ngrams are sliced up is now implemented and I think the class is overall more modular. Unfortunately, during the refactoring process, the functionality of the tool was reduced a bit in that, it can only provided Ngrams for one annotation source at a time. I do not think this is optimal because I foresee the user finding an Ngram of interest and then exploring other annotations it has (e.g. finding a gene-cluster Ngram that occurs a lot then exploring its COG annotations). If you have any suggestions on how the tool can report multiple annotations please let me know :)

Thanks again for the continued advice and suggestions!

**Comment by ekiefl** [comment]
Did you push? run_all_tests.sh is not running:

```

:: Testing anvi-analyze-synteny ignoring genes with no annotation ...

usage: anvi-analyze-synteny [-h] -g GENOMES_STORAGE
                            [--ngram-window-range NGRAM_WINDOW_RANGE]
                            [-o FILE_PATH] [--annotation-source SOURCE NAME]
                            [-p PAN_DB] [-l] [--analyze-unknown-functions]
                            [-G GENOME_NAMES]
anvi-analyze-synteny: error: the following arguments are required: -g/--genomes-storage
```

**Comment by ekiefl** [comment]
It looks really good. I pushed some formatting. Good implementation of the 'separation of functionality' that we talked about.

**Comment by mschecht** [comment]
> Did you push? run_all_tests.sh is not running:

Sorry, @ekiefl just did!

**Comment by ekiefl** [comment]
In `run_all_tests.sh`, each output has:

```
* What do we say to loci that appear to have no coherent synteny patterns...? Not
today! ⚔️
```

**Comment by mschecht** [comment]
> In `run_all_tests.sh`, each output has:
> 
> ```
> * What do we say to loci that appear to have no coherent synteny patterns...? Not
> today! ⚔️
> ```
> 
> Is that expected?

Yes, that is expected. Just a cute message to the user confirming that Ngrams are being formed.

**Comment by ekiefl** [comment]
> Yes, that is expected. Just a cute message to the user confirming that Ngrams are being formed.

Got it

**Comment by mschecht** [comment]
@meren and @ekiefl, I have made a lot of improvements. When you get a chance please take a look.

In summary, the tool now accepts a `genome-storage` and accepts multiple annotation sources (e.g. gene-clusters, annotations, or both). If both annotations are provided, the user must provide an additional parameter, `--ngram-source`, which designates which annotation source will be used to construct the ngrams, while the other annotation source is just to add additional functions to the ngrams. Also, if the ngram is flipped during construction, the accompanied another is too :)

**Comment by meren** [comment]
sounds great to me :)

**Comment by ekiefl** [comment]
Great job! I think you should merge.

**Comment by mschecht** [comment]
Thanks everyone!

**Pull request merged**